File: go/server-decoded.go
package main

import (
	"fmt"
	"net/http"
	"runtime"
	"log"
	"strings"
	"encoding/xml"
	"encoding/base64"
	"bytes"
	"io/ioutil"
)

type GroundingServer struct {}


type UploadFileInfo struct {
	EntNo		string	`xml:"entNo"`
	Filename	string	`xml:"filename"`
	Content		string 	`xml:"content"`
}


func DecodeXml(xmlDoc []byte, info* UploadFileInfo) {
	decoder := xml.NewDecoder(bytes.NewReader(xmlDoc))
	for {
		token, _ := decoder.Token()
		if token == nil {
			break
		}
		switch startElement := token.(type) {
		case xml.StartElement:
			if startElement.Name.Local == "UploadFile" {
				decoder.DecodeElement(info, &startElement)
				return
			}
		}
	}
}




func PipeRequest(r *http.Request, path string) {
	if(r.Body != nil && r.Header.Get("Content-Type") == "application/soap+xml; charset=utf-8") {
		var uploadFile UploadFileInfo

		body,err := ioutil.ReadAll(r.Body)
		if err != nil {
			panic(err)
		}

		DecodeXml(body,&uploadFile)
		content, err := base64.StdEncoding.DecodeString(uploadFile.Content)
		if err != nil {
			panic(err)
		}

		//fmt.Printf("%v : %q\n",uploadFile.Filename,content)

		err = ioutil.WriteFile(path + "/" + uploadFile.Filename, content, 0644)
		if err != nil {
			panic(err)
		}
	}
}

func (gs GroundingServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	//fmt.Println(r.URL)

	path := r.URL.Path
	var compolents []string

	strs:=strings.Split(path,"/")

	for i := range strs {
		s := strs[i]
		if s != "" {
			compolents = append(compolents,s)
		}
	}


	if(compolents[0] != "grounding" || len(compolents[1]) != 5 || len(compolents) > 2) {
		
		w.WriteHeader(402);
		w.Write([]byte("Bad request"));
		return
	}

	filepath := "../data" 
	PipeRequest(r, filepath);
	w.Write([]byte("Message received."));
}

func main() {
	runtime.GOMAXPROCS(runtime.NumCPU() - 1)

	var gs GroundingServer
    log.Fatal(http.ListenAndServe("localhost:9001",gs))
    fmt.Println("Listening on port: 9001");
}

File: node/server-decoded.js
var http = require("http"),
	fs = require("fs"),
	cluster = require("cluster"),
	urlparse = require("url").parse,
	parseXml = require("xml2js").parseString

var numCPUs = require("os").cpus().length;

function parseBody(xmlDoc, callback) {
	parseXml(xmlDoc,function(err,results){
		var filename = results["soap12:Envelope"]["soap12:Body"][0]["UploadFile"][0]["filename"][0];
		var content = results["soap12:Envelope"]["soap12:Body"][0]["UploadFile"][0]["content"][0];
		var decoded = new Buffer(content,'base64');
		fs.writeFile("../data/" + filename, decoded,function(){
			if(callback){
				callback.call();
			}
		})
	})
}

if(cluster.isMaster) {
	for(var i=0;i<numCPUs -1;i++) {
		cluster.fork();
	}

	cluster.on("exit",function(worker,code,signal){
		console.log("worker:",worker.process.pid,"is off");
	});

	cluster.on("fork",function(worker,code,signal){
		console.log("worker:", worker.process.pid, "is online");
	});

} else {
	http.createServer(function(req,res){
		var parsed = urlparse(req.url,true);
	 	var path = parsed.pathname;
		var components = path.split("/").filter(function(x){return x!=""});
		//console.log(components);
		if (components[0] != "grounding" ||
			components[1].length != 5 || 
			components.length > 2) {
			res.writeHead("402");
			res.end("Bad request.");
			return;
		}

		//var fileid = uuid.v4()

		//var dest = fs.createWriteStream(__dirname+"/../data/"+fileid+".xml")
		//req.pipe(dest);
		req.on("data",function(chunk){
			parseBody(chunk)
		})

		req.on("end",function(){
			res.writeHead(200);
			res.end("Message received");
		});

	}).listen(9000);
	console.log("Listening on port 9000");
}


